### PR TITLE
Enable fine-tuning in Deepmil

### PR DESCRIPTION
**Pull request by harshita-s**
This PR contains the following:
- Fine-tuning can be enabled or disabled in Deepmil for PANDA.
- Correct way of initializing HistoSSL encoder.

### CODE REVIEW

**Review by ant0nsc — APPROVED** [review]
Code looks good - but do we have any way of testing it?

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py [hunk 1]
```diff
@@ -41,6 +42,9 @@ def __init__(self, **kwargs: Any) -> None:
             pooling_type=GatedAttentionLayer.__name__,
             # average number of tiles is 56 for PANDA
             encoding_chunk_size=60,
+            cache_mode=CacheMode.MEMORY,
+            precache_location=CacheLocation.SAME,
```
Comment (vale-salvatelli) [inline InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py:46]: if we're fine-tuning what is the advantage of caching(save encoded features) and pre-caching(saving the all dataset up-front)?

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py [hunk 2]
```diff
@@ -103,12 +108,14 @@ def setup(self) -> None:
...
     def get_data_module(self) -> PandaTilesDataModule:
         image_key = PandaTilesDataset.IMAGE_COLUMN
-        transform = Compose(
-            [
-                LoadTilesBatchd(image_key, progress=True),
-                EncodeTilesBatchd(image_key, self.encoder, chunk_size=self.encoding_chunk_size),
-            ]
-        )
+        if self.is_finetune:
+            transform = Compose([LoadTilesBatchd(image_key, progress=True)])
```
Comment (vale-salvatelli) [inline InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py:112]: should we overwrite the cache parameters in this case?

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py (hunk 1 above)
Comment (harshita-s) [reply InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py:46]: In case of finetuning, transform only includes loading tiles and not encoding the tiles. The caching will be of the tiles and not the features

File: Tests/SSL/test_ssl_containers.py [hunk 3]
```diff
@@ -634,7 +634,24 @@ def test_simclr_dataloader_type() -> None:
...
+            assert isinstance(batch[1], torch.Tensor)
+            if i == 1:
+                break
+
+        for i, batch in enumerate(dataloader[SSLDataModuleType.LINEAR_HEAD]):
+            assert isinstance(batch[0], torch.Tensor)
+            assert isinstance(batch[1], torch.Tensor)
+            assert isinstance(batch[2], torch.Tensor)
+            if i == 1:
+                break
```
Comment (vale-salvatelli) [inline Tests/SSL/test_ssl_containers.py:652]: I guess this is an addition from Anton or Max to avoid a separate PR :) ?

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py (hunk 1 above)
Comment (dccastro) [reply InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py:46]: @vale-salvatelli to further clarify what @harshita-s means, here it's caching the loaded tiles into memory, so they don't need to be read from disk every time. IIRC this made an epoch drop from 30min to 5min in Harshita's experiments

File: Tests/SSL/test_ssl_containers.py (hunk 3 above)
Comment (harshita-s) [reply Tests/SSL/test_ssl_containers.py:652]: yes, correct

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py [hunk 4]
```diff
@@ -41,6 +42,9 @@ def __init__(self, **kwargs: Any) -> None:
             pooling_type=GatedAttentionLayer.__name__,
             # average number of tiles is 56 for PANDA
             encoding_chunk_size=60,
+            cache_mode=CacheMode.MEMORY,
+            precache_location=CacheLocation.SAME,
+            batch_size=8,
```
Comment (vale-salvatelli) [inline InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py:47]: Is batch_size 8 the maximum we can handle during fine-tuning? it seems that the set of hyperparameters we will need to use for DeepMIL with fine tune and without will be pretty different. If we want to leave track of this in the form of configs file in the repo one option is to start defining some child classes, for example one for DeepSMILEPandaFrozen and one for DeepSMILEPandaTuning, where in each we overwrite the parameters of this constructor that needs to be changed

Comment (harshita-s) [reply InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py:47]: Thanks. I will remove this line to keep the original batch size. For me, batch size 8 worked but it will require more experimentation to determine the optimal hyperparameters for finetuning. I would not create the new class as of now, but we could revisit this when we are experimenting more with finetuning. In this PR, I just want to enable the fine-tuning and we could optimize hyperparameters in the specific experiment

File: InnerEye/ML/Histopathology/models/deepmil.py [hunk 5]
```diff
@@ -187,8 +191,11 @@ def log_metrics(self,
                 log_on_epoch(self, f'{stage}/{metric_name}', metric_object)
 
     def forward(self, images: Tensor) -> Tuple[Tensor, Tensor]:  # type: ignore
-        with no_grad():
+        if self.is_finetune:
             H = self.encoder(images)                        # N X L x 1 x 1
+        else:
+            with no_grad():
```
Comment (dccastro) [inline InnerEye/ML/Histopathology/models/deepmil.py:197]: As a cleaner alternative to `if/else`, consider using `with torch.set_grad_enabled(self.is_finetune):` ([docs](https://pytorch.org/docs/stable/generated/torch.set_grad_enabled.html)).

File: InnerEye/ML/Histopathology/models/encoders.py [hunk 6]
```diff
@@ -140,4 +140,6 @@ class HistoSSLEncoder(TileEncoder):
     def _get_encoder(self) -> Tuple[Callable, int]:
         resnet18_model = resnet18(pretrained=False)
         histossl_encoder = load_weights_to_model(self.WEIGHTS_URL, resnet18_model)
-        return setup_feature_extractor(histossl_encoder, self.input_dim)  # type: ignore
+        histossl_encoder.fc = torch.nn.Sequential()
+        num_features = 512
```
Comment (dccastro) [inline InnerEye/ML/Histopathology/models/encoders.py:144]: To avoid hard-coding `num_features`, one option is to take `resnet18_model.fc.in_features`

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py [hunk 7]
```diff
@@ -103,12 +108,14 @@ def setup(self) -> None:
```
Comment (vale-salvatelli) [inline InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py:108]: This cuda() could be the reason why the multi-gpu option is not working - discussed in more detail at voice

Comment (vale-salvatelli) [reply InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py:108]: About .eval(). looking at what happens inside the forward pass of deepmil I don't think we should set it to eval mode if we're in the fine-tune case

File: InnerEye/ML/Histopathology/models/deepmil.py [hunk 8]
```diff
@@ -54,7 +54,8 @@ def __init__(self,
                  slide_dataset: SlidesDataset = None,
                  tile_size: int = 224,
                  level: int = 1,
-                 class_names: Optional[List[str]] = None) -> None:
+                 class_names: Optional[List[str]] = None,
+                 is_finetune: Optional[bool] = False) -> None:
```
[the remark this answers is not in the document]
Comment (harshita-s) [reply InnerEye/ML/Histopathology/models/deepmil.py]: Thanks, changed this now

File: InnerEye/ML/Histopathology/models/deepmil.py [hunk 9]
```diff
@@ -187,7 +191,7 @@ def log_metrics(self,
                 log_on_epoch(self, f'{stage}/{metric_name}', metric_object)
 
     def forward(self, images: Tensor) -> Tuple[Tensor, Tensor]:  # type: ignore
-        with no_grad():
+        with set_grad_enabled(self.is_finetune):
             H = self.encoder(images)                        # N X L x 1 x 1
```
Comment (vale-salvatelli) [inline InnerEye/ML/Histopathology/models/deepmil.py:195]: One additional thought: here we should call the function to encode in chunks, otherwise we will be limited in the max_bag_size as we were back at the beginning of our experiments

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py (hunk 4 above)
Comment (harshita-s) [reply InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py]: I have added the hyperparameters for which fine-tuning works on PANDA in the comments, as suggested earlier

File: InnerEye/ML/Histopathology/models/encoders.py (hunk 6 above)
Comment (harshita-s) [reply InnerEye/ML/Histopathology/models/encoders.py]: Thanks, done

File: InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py (hunk 7 above)
Comment (harshita-s) [reply InnerEye/ML/configs/histo_configs/classification/DeepSMILEPanda.py:112]: This is done, thanks

File: InnerEye/ML/Histopathology/models/deepmil.py (hunk 9 above)
Comment (harshita-s) [reply InnerEye/ML/Histopathology/models/deepmil.py:196]: I think doing this is more involved as monai transforms use dataset and dataloaders and it could be a separate PR.

**Review by vale-salvatelli — COMMENTED** [review]
As discussed I think we need to make this more robust:
* makes sure it can run on multi-gpu and doesn't hit memory errors
* adding tests
But I am happy to approve as first version so that more people can help out

**Review by vale-salvatelli — APPROVED** [review]

File: InnerEye/ML/Histopathology/models/encoders.py [hunk 10]
```diff
@@ -139,5 +139,9 @@ class HistoSSLEncoder(TileEncoder):
 
     def _get_encoder(self) -> Tuple[Callable, int]:
         resnet18_model = resnet18(pretrained=False)
+        num_features = resnet18_model.fc.in_features
         histossl_encoder = load_weights_to_model(self.WEIGHTS_URL, resnet18_model)
-        return setup_feature_extractor(histossl_encoder, self.input_dim)  # type: ignore
+        histossl_encoder.fc = torch.nn.Sequential()
+        for param in histossl_encoder.parameters():
+            param.requires_grad = False
+        return histossl_encoder, num_features  # type: ignore
```
Comment (dccastro) [inline InnerEye/ML/Histopathology/models/encoders.py:147]: Minor suggestion to revert these changes in `_get_encoder()`, as `setup_feature_extractor()` now implements the same behaviour.

**Review by dccastro — APPROVED** [review]

**Review by vale-salvatelli — APPROVED** [review]
